src/hooks/useSmartFetch.ts

```diff
@@ -0,0 +1,102 @@
+import { useCallback, useRef } from "react";
+
+import { useAuth, useToasts } from "../providers";
+import { getCookie } from "../utils";
+
+interface ErrorNetwork {
+  status: "network";
+  error: any;
+}
+
+interface ErrorGeneral {
+  status: "unauthorized" | "forbidden" | "network" | "stale" | "unknown";
+  response: Response;
+}
+
+interface Success<T> {
+  status: "ok";
+  response: Response;
+  parsed: T;
+}
+
+export type SmartFetchResult<T> = ErrorGeneral | ErrorNetwork | Success<T>;
+
+export const useSmartFetch = () => {
+  return useSmartFetchUnauthed(useAuth().signOut);
+};
+
+export const useSmartFetchUnauthed = (signOut?: () => Promise<SmartFetchResult<void>>) => {
```

Comment:
When refactoring the useCallback/useMemo, I ended up removing the coupling. I agree it's better decoupled and more explicit if we don't include it in the `smartFetch`, but, in the long run, I would probably end up coupling some of the logic somewhere to avoid having to take care of e.g. network errors [every](https://github.com/tadeaspetak/teleport-challenge/blob/18066fd55ecec2c6be4ec78fe77434ea5c6f7169/src/views/SignIn/index.tsx#L34) [time](https://github.com/tadeaspetak/teleport-challenge/blob/18066fd55ecec2c6be4ec78fe77434ea5c6f7169/src/views/Browse/index.tsx#L39) anything's fetched.